llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -57,197 +57,139 @@ static cl::opt<bool>
-  assert(MI.getNumOperands() == 2 && "Expected only 2 operands in the MI");
-
-  if (!MI.getOperand(1).isReg())
-    return;
-
-  Register DstPhysReg = MI.getOperand(0).getReg();
-  Register SrcReg = MI.getOperand(1).getReg();
-  // We currently have $PhysReg = COPY SrcReg. Check if the SrcReg is virtual
-  // and find the def of the SrcReg to check for %virtreg = COPY $PhysReg.
-  if (ErasableVirtRegToPhysRegCopies.count(SrcReg) ||
-      !isCopyFromPhysReg(MI, DstPhysReg)) {
-    return;
   }
 
-  // If the copy is found and copy propagation will lead to identity copy,
-  // then just store the current $PhysReg = COPY %virtreg MI to be removed
-  // later.
-  ErasableVirtRegToPhysRegCopies[SrcReg] = &MI;
-}
-
-/// If there is an intervening clobber but copy propagation suggests it
-/// comes from the same PhysReg, it can be removed.
-bool AIEPostSelectOptimize::eraseSimilarClobber(MachineInstr &MI,
-                                                MachineInstr &LastVisitedDef) {
-  if (!isCopyFromPhysReg(LastVisitedDef, MI.getOperand(0).getReg()) ||
-      !isCopyFromPhysReg(MI, LastVisitedDef.getOperand(0).getReg()))
+  /// Return true if \p PhysReg has an available copy in a VReg.
+  bool hasVirtualCopy(MCRegister PhysReg) {
+    if (auto It = Copies.find(PhysReg); It != Copies.end())
+      return !It->second.empty();
     return false;
-
-  Register MISrcReg = MI.getOperand(1).getReg();
-  if (ErasableVirtRegToPhysRegCopies.count(MISrcReg) &&
-      ErasableVirtRegToPhysRegCopies[MISrcReg] == &MI) {
-    ErasableVirtRegToPhysRegCopies.erase(MISrcReg);
   }
-  LLVM_DEBUG(dbgs() << "Erasing redundant MI - " << MI);
-  MI.eraseFromParent();
-  return true;
-}
 
-/// Returns true if the register is an Unallocatable physical register.
-/// E.g. Control Registers.
-inline bool
-AIEPostSelectOptimize::isUnallocatablePhysReg(const MachineOperand &Op) {
-  return Op.isReg() && Op.getReg().isPhysical() && Op.isDef() &&
-         !MRI->isAllocatable(Op.getReg());
-}
-
-/// Erase or store possible redundant copies local to a basic block.
-bool AIEPostSelectOptimize::eraseOrStoreRedundantPhysRegDefs(
-    MachineInstr &MI, DenseMap<Register, PhysRegDefInfoStruct> &PhysRegDefs) {
-  // Update the IsUsed boolean of PhysRegDefs to track for any uses
-  // of the Unallocatable Phys Reg.
-  for (auto &[UnallocatablePhysReg, DefInfo] : PhysRegDefs) {
-    if (MI.readsRegister(UnallocatablePhysReg) ||
-        MI.hasRegisterImplicitUseOperand(UnallocatablePhysReg)) {
-      DefInfo.IsUsed = true;
+  /// Returns true if \p MI is a COPY that was eliminated.
+  bool trySimplifyCopy(MachineInstr &MI) {
+    if (!MI.isCopy())
+      return false;
+    auto [DstReg, SrcReg] = MI.getFirst2Regs();
+    if (DstReg.isPhysical() && isCopy(DstReg, SrcReg)) {
+      // Assigning to DstReg a copy of itself. Just remove MI.
+      MI.eraseFromParent();
+      return true;
     }
+    if (SrcReg.isPhysical() && hasVirtualCopy(SrcReg)) {
```

Comment:
Hi @gbossu. Do you think that this part is responsible for the additional spilling? I mean, we are extending the usage of a virt reg here.